solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/CompositeLTRScoringAlgorithm.java

```diff
@@ -0,0 +1,42 @@
+package org.apache.solr.ltr.ranking;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.util.List;
+
+import org.apache.solr.ltr.feature.LTRScoringAlgorithm;
+import org.apache.solr.ltr.util.ModelException;
+import org.apache.solr.ltr.util.NamedParams;
+
+/**
+ * This model introduces the possibility of having composite scoring algorithms.
+ */
+public abstract class CompositeLTRScoringAlgorithm extends LTRScoringAlgorithm {
+
+  protected LTRScoringAlgorithm scoringAlgorithm;
+
+  public CompositeLTRScoringAlgorithm(String name, List<Feature> features,
```

Comment:
Since we're going down the route for Composite scoring algorithms, should we update this to follow the Composite pattern, enabling us to have more flexibility in the future?  Let me know what you think of this.

- Update this class to have a list of scoringAlgorithms instead, and create an an `add(LTRScoringAlgorithm algo)` that appends it to the scoringAlgorithms list.  
- Move `abstract createLTRModel()` into HeuristicBoostedModel, and in HeuristicBoostedModel's constructor call `this.add( createLTRModel(...) )`
- Update HeuristicBoostedModel to use scoringAlgorithms (loop over them, or access the element since you know you only have 1, or something else)

Theoretically, this would make it easy for people to chain ranking models together if they wanted (assuming they are fast enough).  
In some sense, your HeuristicFeatureBoost is like a mini ScoringAlgorithm that you currently `score()` and `explain()`.  If a common interface was shared between them, I could see you being able to `add()` both the boost and model in HeuristicBoostedModel.  The model could then just loop over them all and `sum | product` their scores and explains, without having any special "if (boost exists)" logic.  What are your thoughts on this part about boosts?  I don't want to introduce scope creep, so just something to think about.